### PR TITLE
fix(insights): Fix timezone date issues

### PR DESCRIPTION
**Pull request by timgl**
## Problem

Lots of random assorted issues with timezones.

## Changes

- Fix loading persons if timezones is active
- Fix loading same day for timezones
- We now do not support date_to with a specific time in filters. We will always set date_to to the end of the day (23:59:59) OR to the current time if interval = hour. We don't set a specific time anywhere in the interface, so this shouldn't be an issue.
- In general clean up how we move dates around. Most of that logic should now live in date_to and be consistent 

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

unit tests

### CODE REVIEW

File: posthog/queries/test/test_trends.py [hunk 1]
```diff
@@ -542,12 +542,38 @@ def test_trends_compare(self):
...
+                    "2020-01-02",
+                    "2020-01-03",
+                    "2020-01-04",
+                ],
+            )
 
+            self.assertEqual(
+                response[1]["days"],
+                [
+                    "2019-12-21",
```
Comment (timgl) [inline posthog/queries/test/test_trends.py:562]: We are changing the behaviour of compare here, but 2020-01-04 minus 14 days is 21 so this makes a lot more sense

**Review by neilkakkar — COMMENTED** [review]
I'm a bit confused with how the time handling works now. Sanity check:

1. Everything is stored in Clickhouse in UTC
2. We convert datetimes we get into UTC before querying them in clickhouse
3. .. Except when they have a time component, in which case we assume/coerce its always in UTC, and then don't try and convert these to whichever timezone was set.


Couldn't see it in the tests yet, but assuming the above is true, I'd expect to see timestamps in Clickhouse with starting and ending being something like: '2020-02-03 05:30:00' and '2020-02-05 21:59:59' instead of always starting at 00:00:00 and ending at 23:59:59, to adjust for the timezones 🤔

**Comment by timgl** [comment]
> Everything is stored in Clickhouse in UTC
> We convert datetimes we get into UTC before querying them in clickhouse
> .. Except when they have a time component, in which case we assume/coerce its always in UTC, and then don't try and convert these to whichever timezone was set.

Yep that's right. The canonical best way of working with timezones is to use UTC everywhere, and only really use timezones for displaying to the user

Obviously any bucketing queries here are the exception as the bucketing changes depending on timezone

File: ee/clickhouse/queries/test/__snapshots__/test_trends.ambr [hunk 2]
```diff
@@ -108,8 +108,8 @@
         GROUP BY group_key) groups_0 ON "$group_0" == groups_0.group_key
      WHERE team_id = 2
        AND event = 'sign up'
-       AND timestamp >= toDateTime('2020-01-02 00:00:00', 'UTC')
-       AND timestamp <= toDateTime('2020-01-02 23:59:59', 'UTC')
+       AND timestamp >= toDateTime('2020-01-02 00:00:00')
+       AND timestamp <= toDateTime('2020-01-03 00:00:00')
```
[the remark this answers is not in the document]
Comment (timgl) [reply ee/clickhouse/queries/test/__snapshots__/test_trends.ambr]: Good catch! Fixed, slightly niche bug that only happens in tests

File: ee/clickhouse/queries/test/__snapshots__/test_lifecycle.ambr [hunk 3]
```diff
@@ -421,3 +421,169 @@
...
+              status
+     ORDER BY start_of_period ASC)
+  GROUP BY status
+  '
+---
+# name: TestClickhouseLifecycle.test_timezones.1
+  '
+  WITH 'day' AS selected_period,
+       periods AS
+    (SELECT dateSub(day, number, dateTrunc(selected_period, toDateTime('2020-01-19 23:59:59', 'US/Pacific'))) AS start_of_period
```
Comment (neilkakkar) [inline ee/clickhouse/queries/test/__snapshots__/test_lifecycle.ambr:511]: same q regd timezones in CH

Comment (neilkakkar) [reply ee/clickhouse/queries/test/__snapshots__/test_lifecycle.ambr:511]: oh nevermind, this one is bucketing!

**Comment by timgl** [comment]
@neilkakkar ready for re-review!

**Review by neilkakkar — APPROVED** [review]
I think this mostly makes sense!

It's a tricky PR though, and i'd appreciate @EDsCODE also taking a look for things I might've missed

File: ee/clickhouse/queries/trends/util.py [hunk 4]
```diff
@@ -73,9 +73,7 @@ def get_active_user_params(filter: Union[Filter, PathFilter], entity: Entity, te
     diff = timedelta(days=7) if entity.math == WEEKLY_ACTIVE else timedelta(days=30)
     if filter.date_from:
         params.update(
-            {
-                "parsed_date_from_prev_range": f"AND timestamp >= '{format_ch_timestamp(filter.date_from - diff, filter)}'"
-            }
+            {"parsed_date_from_prev_range": f"AND timestamp >= '{format_ch_timestamp(filter.date_from - diff)}'"}
```
Comment (EDsCODE) [inline ee/clickhouse/queries/trends/util.py:76]: Is this supposed to need the timezone arg?

File: posthog/queries/util.py [hunk 5]
```diff
@@ -23,44 +23,51 @@ def parse_timestamps(filter: FilterType, team: Team, table: str = "") -> Tuple[s
...
+    )
 
     return date_from or "", date_to or "", params
 
 
-def format_ch_timestamp(
-    timestamp: datetime, filter, default_hour_min: str = " 00:00:00", timezone: Optional[str] = None
-):
-    if timezone:
+def format_ch_timestamp(timestamp: datetime, convert_to_timezone: Optional[str] = None):
```
Comment (EDsCODE) [inline posthog/queries/util.py:61]: Was the timezone issue double conversion? As in, the timezone is handled here and then handled once again in `toDateTime`

**Review by EDsCODE — COMMENTED** [review]
Some clarification on what's changing

File: posthog/queries/util.py (hunk 5 above)
Comment (timgl) [reply posthog/queries/util.py:61]: Yeah partly, when I did it before it was inconsistent which led to issues. Now we'll just always do it here.

File: ee/clickhouse/queries/trends/util.py (hunk 4 above)
Comment (timgl) [reply ee/clickhouse/queries/trends/util.py:76]: it explicitly does not. I made the decision that when we pass through those people URLs all of the dates should be in UTC for consistency. In general when working with timezones you really only want toto use timezones for user display, not for moving times around. see https://github.com/PostHog/posthog/pull/9678/files/98f36e23df1acb96128c75cdd01bb6d8a1837290#diff-cf067174aadfdcda82472c96d1e1278a8b8884f3ec17a1869b0e0e26ce586552R145

**Pull request merged**